platform-sdk/swirlds-merkledb/src/main/java/com/swirlds/merkledb/collections/LongListDisk.java

```diff
@@ -157,57 +161,52 @@ protected void onEmptyOrAbsentSourceFile(final Path path) throws IOException {
-            // writing the full chunk, all values before minValidIndexInChunk are zeroes
-            MerkleDbFileUtils.completelyWrite(tempFileCHannel, transferBuffer, 0);
-            chunkList.set(firstChunkWithDataIndex, 0L);
-
-            // copy everything except for the first chunk and the last chunk
-            final int numberOfFullChunks = totalNumberOfChunks - firstChunkWithDataIndex - 2;
-            if (numberOfFullChunks > 0) {
-                final long bytesToTransfer = (long) numberOfFullChunks * memoryChunkSize;
-                final long bytesTransferred = MerkleDbFileUtils.completelyTransferFrom(
-                        tempFileCHannel, fileChannel, memoryChunkSize, bytesToTransfer);
-                if (bytesTransferred != bytesToTransfer) {
-                    throw new IOException("Failed to read long list chunks, expected=" + bytesToTransfer + " actual="
-                            + bytesTransferred);
-                }
+            this.tempFileChannel = rf.getChannel();
+            super.readBodyFromFileChannelOnInit(sourceFileName, fileChannel);
+        } finally {
+            if (this.tempFileChannel != null && this.tempFileChannel.isOpen()) {
+                this.tempFileChannel.close();
             }
+        }
+    }
 
-            // copy the last chunk
-            transferBuffer.clear();
-            MerkleDbFileUtils.completelyRead(fileChannel, transferBuffer);
-            transferBuffer.flip();
-            MerkleDbFileUtils.completelyWrite(
-                    tempFileCHannel, transferBuffer, (long) lastChunkWithDataIndex * memoryChunkSize);
+    /** {@inheritDoc} */
+    @Override
+    protected Long readChunkData(FileChannel fileChannel, int chunkIndex, int startIndex, int endIndex)
+            throws IOException {
+        final ByteBuffer transferBuffer = initOrGetTransferBuffer();
+        fillBufferWithZeroes(transferBuffer);
 
-            for (int i = firstChunkWithDataIndex + 1; i < totalNumberOfChunks; i++) {
-                chunkList.set(i, (long) (i - firstChunkWithDataIndex) * memoryChunkSize);
-            }
+        readDataIntoBuffer(fileChannel, chunkIndex, startIndex, endIndex, transferBuffer);
+
+        final int firstChunkWithDataIndex = toIntExact(minValidIndex.get() / numLongsPerChunk);
+        return (long) (chunkIndex - firstChunkWithDataIndex) * memoryChunkSize;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    protected void setChunk(int chunkIndex, Long chunk, int startIndex, int endIndex) throws IOException {
+        final ByteBuffer transferBuffer = TRANSFER_BUFFER_THREAD_LOCAL.get();
+
+        int startOffset = startIndex * Long.BYTES;
+        int endOffset = (endIndex + 1) * Long.BYTES;
+
+        transferBuffer.position(startOffset);
+        transferBuffer.limit(endOffset);
+
+        int bytesToWrite = endOffset - startOffset;
+        long bytesWritten = MerkleDbFileUtils.completelyWrite(tempFileChannel, transferBuffer, chunk + startOffset);
```

Comment:
This implies `setChunk()` is only used during initialization, from `readBodyFromFileChannelOnInit()`, otherwise `tempFileChannel` would be null. Very error prone.